### ISSUE TITLE
VATEUD Core API Integration

### ISSUE THREAD

**Issue by blt950** [Vatsim-Scandinavia/controlcenter#758]
We'd like to integrate our task system with VATEUD Core API. We've participated in testing of this API and it's going to be a great leap to automate paper work Training Director does today.

To implement:
- [x] Theory test related requests [facility/training/exams](https://api-core.vateud.net/#/facility/training/exams)
- [x] Solo endorsements [facility/endorsements](https://api-core.vateud.net/#/facility/endorsements)
- [x] Tier 1/2 Endorsements [facility/endorsements](https://api-core.vateud.net/#/facility/endorsements)
- [x] Rating upgrades [facility/user](https://api-core.vateud.net/#/facility/user)
- [x] Mentor list [facility/training](https://api-core.vateud.net/#/facility/training)
- [x] Examiner list [facility/training](https://api-core.vateud.net/#/facility/training)

To ensure requests for trainings like "S1+S2" requests are requested correctly 
- [x] If training consists of multiple ratings, give task requester an option to choose rating

Other
- [x] Add related exam tasks in a own card at training view

⚠️ This changes requires definition of endorsement types in database

**Comment by bedrich-schindler** [comment]
First of all, thank you for all the work you have done! The amount of work done is insane! Thanks a lot!

I have a question about further development. I have check Core API (https://api-core.vateud.net/#/facility%2Ftraining/get_staff) and there many interesting endpoints that are not implemented in prepared PR, so I just would like to know about your plans @thor @blt950.

Q1) Now, it is required to add user to Core prior to CC if am I right. Do you plan to implement some integration with `/roster ` endpoints? So that we can add/remove users directly from CC using button/or automatically using cron that would sync that data between CC and Core? Such integration would be great for all VACCc that are upgrading CC and that are using CC as well. I am not sure how to use it now if we need to add users to Core manually.

Q2) It is connected to Q1. Do you plan to setup some cron that would sync the rest of the data (endorsements, mentors, examiners, exams)? It would simply get data from Core, compare it to CC data and perform automated calls that would sync that data. If yes, it would be interesting to create log into the portal so that we can see that something get synced from Core to CC and vice versa. Such implementation would solve initial migration to Core and will solve out-of-sync data. It should not occur, but you know - users ... :D 

Q3 ) `facility/visitiors` endpoints seem quite important and such integration with CC would be great. For example: I can imagine when new pending visiting request is created in Core, Portal would sync with Core, it would check whether the user is already logged in CC and if not, it would send email to the user to log in to the CC so that we can process such request within CC. It should inform DIR and TD as well. As Core API does not provide email of users that have not log in yet, it would require VATSIM API to obtain it.

There are few more endpoints, but to be honest, events does not need to be implemented in Core as Event department never logs in to our CC. And I am not sure what about exams and rating upgrades.

**Comment by blt950** [comment]
Hi @bedrich-schindler

First of all, all endpoints except of visitors, roster and events are implemented. We've as of today no plans to implement the rest.

**Q1**
As of today, you need all users who should be granted something in Core to have logged into core in advance. Mufassil who develop core has informed he's working on a update so all EUD members will be available in Core without them logging in soon. Hence this PR #789 is on hold, as asking 64 mentors (in VATSCA) to login to just assign them as mentor will be tedious and impossible.

We've no plans on integrating rosters, as CC will be operating as the source of truth and it has to stay that way to respect the privacy first principle. Meaning: A user will need to login into CC to be shows in the CC's roster, this is because that's the only way you make sure they've accepted your privacy policy prior to login (at least if you use Handover), and a logged in user is also then subject to automations that need to be approved by them explicitly.

Oh and, it's more subdivision/division agnostic to use the VATSIM API roster instead of VATEUD's. Then technically it's only one integration and why fix something that ain't broken. (Except of VATSIM Tech using weeks to issue a key recently, but that's not relevant here)

tl;dr: EUD will fix the roster, hence this PR is on hold until it's fixed so this will not be a problem anymore.

**Q2**
Again, this PR is created that CC is source of truth. We could of course still have some sync functionality, but it's not planned and outside of the scope for us. Feel free to contribute a sync functionality if it's of high importance for your vACC. The purpose of this integration is to make CC work for mentors and training staff, and training directors and directors will at the end of day need to enter the EUD portal to do the smaller tasks that was not calculated as valuable to spend development time for in this integration.

**Q3**
This is out of scope. Mostly also because each vACC's visiting application process is different and at least for our (VATSCA) case we don't get _that_ many visitor requests to justify the development time for this.

_Thank you for your questions. While this integration doesn't mark the end of developing it further, there's no current plans to do so (except bug fixes) for the simple reason that I'm the only active developer and have already a big backlog of other things to be solved, so I need to be selective. So we'll be dependent on contributions for further development of this integration._

**Comment by blt950** [comment]
Update: So apparently a roster sync will be required by EUD. So I'll add that as a daily cron job. It'll though only sync one way, CC is source of truth as earlier described.
